s3box_review/s3box.go

```diff
@@ -0,0 +1,243 @@
+// be pulled from your environment.
+type NewS3BoxOptions struct {
+	// S3Bucket is the destination s3 bucket.
+	// This is required.
+	S3Bucket string
+
+	// AWSKey is the AWS ACCESS KEY ID.
+	// By default grabs from your environment.
+	AWSKey string
+
+	// AWSPassword is the AWS SECRET ACCESS KEY.
+	// By default grabs from your environment.
+	AWSPassword string
+
+	// AWSToken is the AWS SESSION TOKEN.
+	// By default grabs from your environment.
+	AWSToken string
+
+	// BufferSize is the maximum size of data we buffer internally
+	// before creating an s3 file.
+	// This is optional and defaults to 10MB.
+	BufferSize int
+}
+
+// NewS3Box creates a new S3Box given the input options.
+// Errors occur if there's an invalid input or if there's difficulty setting up an s3 connection.
+func NewS3Box(options NewS3BoxOptions) (*S3Box, error) {
+	// Check for required inputs and a valid destination config
+	if options.S3Bucket == "" {
+		return nil, ErrS3BucketRequired
+	}
+
+	bufferSize := DefaultBufferSize
+	if options.BufferSize > 0 {
+		bufferSize = options.BufferSize
+	}
+
+	// Setup s3 handler and aws configuration. If no creds are explicitly provided, they'll be grabbed from the environment.
+	region, err := getRegionForBucket(options.S3Bucket)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get AWS region for bucket %s: (%s)", options.S3Bucket, err)
+	}
+
+	// If AWS creds were provided use those, otherwise grab them from your environment
+	var awsCreds *credentials.Credentials
+	if options.AWSKey == "" && options.AWSPassword == "" && options.AWSToken == "" {
+		awsCreds = credentials.NewEnvCredentials()
+	} else {
+		awsCreds = credentials.NewStaticCredentials(options.AWSKey, options.AWSPassword, options.AWSToken)
+	}
+	awsConfig := aws.NewConfig().WithRegion(region).WithS3ForcePathStyle(true).WithCredentials(awsCreds)
+	awsSession := session.New()
+
+	return &S3Box{
+		s3Bucket:   options.S3Bucket,
+		timestamp:  time.Now(),
+		s3Handler:  s3.New(awsSession, awsConfig),
+		bufferSize: bufferSize,
+	}, nil
+}
+
+// NextBox gives you a new box, forgetting everything about previously packaged data
+func (sb *S3Box) NextBox() {
+	sb.Lock()
+	sb.timestamp = time.Now()
+	sb.fileNumber = 0
+	sb.fileLocations = []string{}
+	sb.bufferedData = []byte{}
+	sb.isSealed = false
+	sb.Unlock()
+}
+
+// Pack writes bytes into a buffer. Once that buffer hits capacity, the data is output to s3.
+// Any error will leave the buffer unmodified.
+func (sb *S3Box) Pack(data []byte) error {
+	if sb.isSealed {
+		return ErrBoxIsSealed
+	}
+
+	// If the bytes aren't in JSON format, return an error
+	var tempMap map[string]interface{}
+	if err := json.Unmarshal(data, &tempMap); err != nil {
+		return ErrInvalidJSONInput
+	}
+
+	sb.Lock()
+	oldBuffer := sb.bufferedData // If write fails, keep buffered data unchanged
+	data = append(data, '\n')    // Append a new line for text-editor readability
+	sb.bufferedData = append(sb.bufferedData, data...)
+	sb.Unlock()
```

Comment:
I'm thinking of changing this to
```
defer sb.Unlock()
```
and keep the entire pack locked to one at a time.